File: src/aoc2021/day04.clj
(ns aoc2021.day04
  (:require [clojure.java.io :as io]
            [clojure.string :as s]))

(comment
 "Day 4: Giant Squid")

(def sample-input
  (-> "../resources/day04_ex.txt" io/resource io/reader line-seq))

(def puzzle-input
  (-> "../resources/day04.txt" io/resource io/reader line-seq))

(defn string->row
  [s]
  (let [split (-> s s/trim (s/split #"\s+"))]
    (mapv read-string split)))

(defn init
  [data]
  (let [parse-board (fn [raw-board] (mapv string->row raw-board))
        raw-calls   (-> (first data) (s/split  #","))
        boards      (->> (drop 2 data)
                         (remove empty?)
                         (partition 5)
                         (mapv parse-board))]
    {:called   '() ; These are lists so that "conj" prepends rather than appends
     :winners  '()
     :uncalled (map read-string raw-calls)
     :boards   boards}))

;; Define the terminating condition

(defn cols
  "Reads rows column-wise."
  [board]
  (apply mapv vector board))

(defn has-bingo?
  "Returns true if a board has a row or column bingo."
  [called board]
  (let [called-set (set called)]
    ;; All numbers of some row in a given board have been called
    (or (some (fn [row] (every? called-set row))
              board)
        ;; ALl numbers of some col in a given board have been called
        (some (fn [col] (every? called-set col))
              (cols board)))))

;; Implement the gameplay mechanism

(defn score
  "Scores a winning bingo board (when used on the turn where it wins)."
  [called board]
  (let [sum-unmarked (->> board
                          (reduce into [])
                          (remove (set called))
                          (apply +))]
    (* sum-unmarked (first called))))

(defn play-bingo
  [{:keys [called uncalled boards winners] :as state} &
   {:keys [last-winner] :or {last-winner false} :as mode}]
  (cond
    (and (not last-winner) (not-empty winners)) (first winners)  ;; Part 1
    (and last-winner (empty? boards))           (first winners)  ;; Part 2
    :else (let [new-winner (first (filter (partial has-bingo? called) boards))
                remaining  (remove (partial has-bingo? called) boards)]
            (recur (cond-> state
                     new-winner (update :winners  conj (score called new-winner))
                     :always    (update :called   conj (first uncalled))
                     :always    (update :uncalled rest)
                     :always    (assoc  :boards   remaining))
                   mode))))

(time
 (let [sample (init sample-input) puzzle (init puzzle-input)]
   (println "Day 4: Giant Squid")
   (println "[Part 1] Sample:" (play-bingo sample) "Puzzle:" (play-bingo puzzle))
   (println "[Part 2] Sample:" (play-bingo sample :last-winner true) "Puzzle:" (play-bingo puzzle :last-winner true))))

File: py/day04.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-

import re
from itertools import chain

# https://stackoverflow.com/questions/6560354/how-would-i-create-a-custom-list-class-in-python
class Board(list):
    def __init__(self, *args, **kwargs):
        super(Board, self).__init__(args[0])

    def cols(self):
        return [[row[i] for row in self] for i in range(len(self[0]))]

    def has_bingo(self, called):
        if any([all([x in called for x in row]) for row in self]) or \
           any([all([x in called for x in col]) for col in self.cols()]):
           return True
        return False

    def score(self, called):
        uncalled = [x for x in chain.from_iterable(self) if x not in called]
        return sum(uncalled) * called[-1]


def parse(path):
    calls, boards = [], []
    cur_board = Board([])
    with open(path) as f:
        for line in f.readlines():
            if ',' in line:
                call_list = [int(x) for x in line.split(',')]
            elif re.search(r'\d+', line):
                cur_board.append([int(x) for x in re.findall(r'\d+', line)])
                if len(cur_board) == 5:
                    boards.append(cur_board)
                    cur_board = Board([])
    return {'call_list': call_list, 'boards': boards}


sample_input = parse('../resources/day04_ex.txt')
puzzle_input = parse('../resources/day04.txt')


def part_1(data):
    called = []
    for call in data['call_list']:
        called.append(call)
        for board in data['boards']:
            if board.has_bingo(called):
                return board.score(called)
    return None


def part_2(data):
    called, winning_scores, remaining = [], [], data['boards']
    for call in data['call_list']:
        called.append(call)
        winning_scores.extend([board.score(called) for board in remaining if board.has_bingo(called)])
        remaining = [board for board in remaining if not board.has_bingo(called)]
    return winning_scores[-1]


print("Day 4\n"
      "[Part 1] Example: %d, Puzzle: %d\n"
      "[Part 2] Example: %d, Puzzle: %d"
      %(part_1(sample_input), part_1(puzzle_input),
        part_2(sample_input), part_2(puzzle_input)))
